hibernate/cache-commons/src/main/java/org/infinispan/hibernate/cache/commons/util/VersionedEntry.java

```diff
@@ -56,27 +71,105 @@ public String toString() {
 		return sb.toString();
 	}
 
-	private static class ExcludeEmptyFilter implements KeyValueFilter<Object, Object>, Converter<Object, Object, Object> {
-      private static final Log log = LogFactory.getLog(ExcludeEmptyFilter.class);
+	@Override
+	public Void apply(EntryView.ReadWriteEntryView<Object, Object> view) {
+		if (trace) {
+			log.tracef("Applying %s to %s", this, view.find().orElse(null));
+		}
+		if (version == null) {
+			// eviction or post-commit removal: we'll store it with given timestamp
+			view.set(this, region.getExpiringMetaParam());
+			return null;
+		}
 
-		@Override
-		public boolean accept(Object key, Object value, Metadata metadata) {
-         log.tracef("Filter check for %s", value);
-			if (value instanceof VersionedEntry) {
-				return ((VersionedEntry) value).getValue() != null;
-			}
-			return true;
+		Object oldValue = view.find().orElse(null);
+		Object oldVersion = null;
+		long oldTimestamp = Long.MIN_VALUE;
+		if (oldValue instanceof VersionedEntry) {
+			VersionedEntry oldVersionedEntry = (VersionedEntry) oldValue;
+			oldVersion = oldVersionedEntry.version;
+			oldTimestamp = oldVersionedEntry.timestamp;
+			oldValue = oldVersionedEntry.value;
+		} else {
+			oldVersion = findVersion(oldValue);
 		}
 
-		@Override
-		public Object convert(Object key, Object value, Metadata metadata) {
-			if (value instanceof VersionedEntry) {
-				return ((VersionedEntry) value).getValue();
+		if (oldVersion == null) {
+			assert oldValue == null || oldTimestamp != Long.MIN_VALUE : oldValue;
+			if (timestamp <= oldTimestamp) {
```

Comment:
~I'm not sure this is a good use of asserts. You're doing an `if` call which only matters if the asserts are enabled. I'd favour asserts not to be used and used normal exception reporting for consistency with the rest of the codebase.~

Actually, I see the if matters from a normal logic perspective. We hardly use `assert` throughout our code base. It worries me to start using them without some clear guidelines on when to use it, as well as some documentation for our users to enable them?

Reply:
Users are not supposed to enable them, the asserts are intended to check for unexpected states in the testsuite.